src/nl/hannahsten/texifyidea/index/LatexProjectStructure.kt

```diff
@@ -0,0 +1,250 @@
+package nl.hannahsten.texifyidea.index
+
+import com.intellij.openapi.project.DumbService
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.openapi.vfs.findPsiFile
+import com.intellij.psi.PsiFile
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.psi.util.CachedValueProvider.Result
+import com.intellij.psi.util.CachedValuesManager
+import nl.hannahsten.texifyidea.psi.LatexParameterText
+import nl.hannahsten.texifyidea.util.TexifyProjectCacheService
+import nl.hannahsten.texifyidea.util.files.LatexPackageLocation
+import nl.hannahsten.texifyidea.util.getBibtexRunConfigurations
+import nl.hannahsten.texifyidea.util.getLatexRunConfigurations
+import nl.hannahsten.texifyidea.util.getTexinputsPaths
+import nl.hannahsten.texifyidea.util.magic.CommandMagic
+import nl.hannahsten.texifyidea.util.parser.collectSubtreeTyped
+import java.nio.file.InvalidPathException
+import java.nio.file.Path
+import kotlin.io.path.Path
+import kotlin.io.path.extension
+import kotlin.io.path.invariantSeparatorsPathString
+import kotlin.io.path.pathString
+
+// May be modified in the future
+/**
+ * A fileset is a set of files that are related to each other, e.g. a main file and its included files (including itself).
+ *
+ * When resolving subfiles, we must use the relative path from the root file rather than the file that contains the input command.
+ */
+data class Fileset(
+    val files: Set<VirtualFile>,
+    val root: VirtualFile,
+)
+
+data class FilesetData(
+    val filesets: Set<Fileset>,
+    val relatedFiles: Set<VirtualFile>,
+    val filesetScope: GlobalSearchScope,
+)
+
+data class LatexProjectFilesets(
+    val filesets: Set<Fileset>,
+    val mapping: Map<VirtualFile, FilesetData>,
+) {
+    fun getData(project: Project, file: VirtualFile): FilesetData? {
+        return mapping[file]
+    }
+}
+
+fun pathOrNull(pathText: String?): Path? {
+    if (pathText.isNullOrBlank()) return null
+    return try {
+        Path(pathText)
+    }
+    catch (e: InvalidPathException) {
+        null
+    }
+}
+
+object LatexProjectStructure {
+
+    const val CACHE_EXPIRATION_IN_MS = 5000L // 5 seconds
+
+    private class BuildFilesetPreparation(
+        val project: Project,
+        val texInputPaths: Set<Path>,
+        val bibInputPaths: Set<Path>,
+    )
+
+    /**
+     *
+     */
+    private fun appendReferredFile(
```

Comment:
I admit that the `isBuildingFileset` parameter for `InputFileReference#resolve` is not very nice, but now it looks like you have just copied over some code that is relevant for the file set, which leads to duplicate code, which is also not ideal. For example, `\tikzfig` support now has gone missing. Can this be done in a better way, which minimizes duplicate code?